pkg/terra/client/client.go

```diff
@@ -0,0 +1,333 @@
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	"github.com/smartcontractkit/terra.go/tx"
+	rpchttp "github.com/tendermint/tendermint/rpc/client/http"
+
+	tmtypes "github.com/cosmos/cosmos-sdk/client/grpc/tmservice"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	txtypes "github.com/cosmos/cosmos-sdk/types/tx"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/smartcontractkit/terra.go/msg"
+	wasmtypes "github.com/terra-money/core/x/wasm/types"
+
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+
+	cosmosclient "github.com/cosmos/cosmos-sdk/client"
+	"github.com/pkg/errors"
+	"github.com/terra-money/core/app"
+
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+
+	"github.com/smartcontractkit/terra.go/key"
+)
+
+//go:generate mockery --name ReaderWriter --output ./mocks/
+type ReaderWriter interface {
+	Writer
+	Reader
+}
+
+// Only depends on the cosmos sdk types.
+type Reader interface {
+	GasPrice() sdk.DecCoin
+	Account(address sdk.AccAddress) (uint64, uint64, error)
+	ContractStore(contractAddress string, queryMsg []byte) ([]byte, error)
+	TxsEvents(events []string) (*txtypes.GetTxsEventResponse, error)
+	Tx(hash string) (*txtypes.GetTxResponse, error)
+	LatestBlock() (*tmtypes.GetLatestBlockResponse, error)
+	BlockByHeight(height int64) (*tmtypes.GetBlockByHeightResponse, error)
+	Balance(addr sdk.AccAddress, denom string) (*sdk.Coin, error)
+}
+
+type Writer interface {
+	// Assumes all msgs are for the same from address.
+	// We may want to support multiple from addresses + signers if a use case arises.
+	SignAndBroadcast(msgs []sdk.Msg, accountNum uint64, sequence uint64, gasPrice sdk.DecCoin, signer key.PrivKey, mode txtypes.BroadcastMode) (*txtypes.BroadcastTxResponse, error)
+	Broadcast(txBytes []byte, mode txtypes.BroadcastMode) (*txtypes.BroadcastTxResponse, error)
+	Simulate(txBytes []byte) (*txtypes.SimulateResponse, error)
+	SimulateUnsigned(msgs []sdk.Msg, sequence uint64) (*txtypes.SimulateResponse, error)
+	CreateAndSign(msgs []sdk.Msg, account uint64, sequence uint64, gasLimit uint64, gasPrice sdk.DecCoin, signer key.PrivKey, timeoutHeight uint64) ([]byte, error)
+}
+
+var _ ReaderWriter = (*Client)(nil)
+
+const (
+	DefaultTimeout = 5
+)
+
+type Logger interface {
+	Infof(format string, values ...interface{})
+	Warnf(format string, values ...interface{})
+	Errorf(format string, values ...interface{})
+}
+
+type Client struct {
+	codec *codec.LegacyAmino
+
+	fallbackGasPrice   sdk.Dec
+	gasLimitMultiplier sdk.Dec
+	fcdURL             string
+	chainID            string
+	clientCtx          cosmosclient.Context
+	sc                 txtypes.ServiceClient
+	ac                 authtypes.QueryClient
+	wc                 wasmtypes.QueryClient
+	bc                 banktypes.QueryClient
+	tmc                tmtypes.ServiceClient
+
+	// Timeout for node interactions
+	timeout time.Duration
+
+	log Logger
+}
+
+func NewClient(chainID string,
+	fallbackGasPrice string,
```

Comment:
Feels to me like this only needs to be a param of the `GasPrice()` method. It's not used anywhere else.